File: knowledge_graph/validate_graph_contract.py
#!/usr/bin/env python3
from __future__ import annotations

import csv
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent
IMPORT = ROOT / "import"
SCHEMA = ROOT / "graph_schema.json"


def main() -> int:
    contract = json.loads(SCHEMA.read_text(encoding="utf-8"))
    checks = {}
    counts = {}
    errors = []
    csv_rows = {}
    for filename, required in contract["required_files"].items():
        path = IMPORT / filename
        checks[f"exists:{filename}"] = path.exists()
        if not path.exists():
            errors.append(f"missing file: {filename}")
            continue
        with path.open(encoding="utf-8-sig", newline="") as handle:
            reader = csv.DictReader(handle)
            header = reader.fieldnames or []
            file_rows = list(reader)
            count = len(file_rows)
            csv_rows[filename] = file_rows
        counts[filename] = count
        missing = [column for column in required if column not in header]
        checks[f"columns:{filename}"] = not missing
        checks[f"nonempty:{filename}"] = count > 0
        if missing: errors.append(f"{filename} missing columns: {', '.join(missing)}")
        if count == 0: errors.append(f"empty file: {filename}")

    node_files = {
        "nodes_job.csv": "job_id:ID",
        "nodes_skill.csv": "skill_id:ID",
        "nodes_job_cluster.csv": "cluster_id:ID",
    }
    node_ids = {}
    for filename, id_column in node_files.items():
        ids = [row.get(id_column, "").strip() for row in csv_rows.get(filename, [])]
        unique_ids = set(ids)
        valid = bool(ids) and "" not in unique_ids and len(ids) == len(unique_ids)
        checks[f"unique_ids:{filename}"] = valid
        node_ids[filename] = unique_ids
        if not valid:
            errors.append(f"{filename} has blank or duplicate {id_column} values")

    relation_contracts = {
        "rel_job_requires_skill.csv": ("nodes_job.csv", "nodes_skill.csv"),
        "rel_job_belongs_cluster.csv": ("nodes_job.csv", "nodes_job_cluster.csv"),
        "rel_skill_parent.csv": ("nodes_skill.csv", "nodes_skill.csv"),
    }
    for filename, (start_file, end_file) in relation_contracts.items():
        relation_rows = csv_rows.get(filename, [])
        missing_start = [row.get(":START_ID", "") for row in relation_rows if row.get(":START_ID", "") not in node_ids.get(start_file, set())]
        missing_end = [row.get(":END_ID", "") for row in relation_rows if row.get(":END_ID", "") not in node_ids.get(end_file, set())]
        pairs = [(row.get(":START_ID", ""), row.get(":END_ID", "")) for row in relation_rows]
        checks[f"valid_endpoints:{filename}"] = not missing_start and not missing_end
        checks[f"unique_pairs:{filename}"] = len(pairs) == len(set(pairs))
        if missing_start or missing_end:
            errors.append(f"{filename} has {len(missing_start)} missing start and {len(missing_end)} missing end nodes")
        if len(pairs) != len(set(pairs)):
            errors.append(f"{filename} has duplicate start/end pairs")

    result = {
        "contract_version": contract["version"],
        "passed": not errors,
        "checks": checks,
        "counts": counts,
        "errors": errors,
    }
    report = IMPORT / "graph_contract_report.json"
    report.write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps(result, ensure_ascii=False, indent=2))
    return 0 if result["passed"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: knowledge_graph/validate-graph-contract.js
import { readFileSync, writeFileSync, existsSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const root = dirname(fileURLToPath(import.meta.url))
const importDir = join(root, 'import')
const contract = JSON.parse(readFileSync(join(root, 'graph_schema.json'), 'utf8'))
const checks = {}
const counts = {}
const errors = []

function parseCsvLine(line) {
  const cells = []
  let value = ''
  let quoted = false
  for (let i = 0; i < line.length; i++) {
    const char = line[i]
    if (char === '"') {
      if (quoted && line[i + 1] === '"') { value += '"'; i++ }
      else quoted = !quoted
    } else if (char === ',' && !quoted) {
      cells.push(value)
      value = ''
    } else value += char
  }
  cells.push(value)
  return cells
}

for (const [filename, required] of Object.entries(contract.required_files)) {
  const path = join(importDir, filename)
  const exists = existsSync(path)
  checks[`exists:${filename}`] = exists
  if (!exists) {
    errors.push(`missing file: ${filename}`)
    continue
  }
  const lines = readFileSync(path, 'utf8').replace(/^\uFEFF/, '').split(/\r?\n/).filter(Boolean)
  const header = parseCsvLine(lines[0] || '')
  const count = Math.max(lines.length - 1, 0)
  counts[filename] = count
  const missing = required.filter(column => !header.includes(column))
  checks[`columns:${filename}`] = missing.length === 0
  checks[`nonempty:${filename}`] = count > 0
  if (missing.length) errors.push(`${filename} missing columns: ${missing.join(', ')}`)
  if (!count) errors.push(`empty file: ${filename}`)
}

const result = {
  contract_version: contract.version,
  passed: errors.length === 0,
  checks,
  counts,
  errors,
}
writeFileSync(join(importDir, 'graph_contract_report.json'), JSON.stringify(result, null, 2), 'utf8')
console.log(JSON.stringify(result, null, 2))
if (!result.passed) process.exitCode = 1
